File: apps/android/feature/exercises/domain/src/main/java/com/scribblefit/feature/exercises/domain/usecase/RemoveExerciseUseCase.kt
package com.scribblefit.feature.exercises.domain.usecase

import com.scribblefit.core.common.runCatchingWithCancellation
import com.scribblefit.feature.exercises.domain.ExerciseRepository
import com.scribblefit.feature.scribble.domain.ScribbleRepository
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.withContext

class RemoveExerciseUseCase(
    private val repository: ExerciseRepository,
    private val scribbleRepository: ScribbleRepository,
    private val coroutineDispatcher: CoroutineDispatcher
) {
    suspend operator fun invoke(exerciseId: Long): Result<Unit> = withContext(coroutineDispatcher) {
        runCatchingWithCancellation {
            val exercise = repository.getExerciseById(exerciseId) ?: return@runCatchingWithCancellation
            val scribbleId = exercise.scribbleId
            
            repository.deleteExercise(exerciseId)
            
            val remainingExercises = repository.getExercisesForScribble(scribbleId)
            if (remainingExercises.isEmpty()) {
                scribbleRepository.deleteScribble(scribbleId)
            }
        }
    }
}

File: apps/ios/ScribbleFit/LocalPackages/Features/FeatureExercises/Sources/FeatureExercises/Domain/UseCases/RemoveExerciseUseCase.swift
import Foundation
import CoreModel

@MainActor
public final class RemoveExerciseUseCase {
    private let exerciseRepository: ExerciseRepository
    private let scribbleRepository: ScribbleRepository
    
    public init(exerciseRepository: ExerciseRepository, scribbleRepository: ScribbleRepository) {
        self.exerciseRepository = exerciseRepository
        self.scribbleRepository = scribbleRepository
    }
    
    public func execute(id: UUID) async throws {
        guard let exercise = try await exerciseRepository.getExercise(id: id) else { return }
        let scribbleId = exercise.scribbleId
        
        try await exerciseRepository.deleteExercise(id: id)
        
        if let sId = scribbleId {
            if let scribble = try await scribbleRepository.getScribble(id: sId) {
                if scribble.exercises.isEmpty {
                    try await scribbleRepository.deleteScribble(id: sId)
                }
            }
        }
    }
}
